node/node-sdk.md

```diff
@@ -0,0 +1,614 @@
+  console.log(greeting);
+}
+
+export const workflow: Example = { main };
+```
+
+##### Activites with no type definitions
+```ts
+const Greet = (name: string) => string;
+
+const greet = Context.configure<Greet>('greet', { // ActivityOptions
+  type: 'remote', // 'local' is also valid
+  taskQueue: 'greeter',
+  startToCloseTimeout: '1 hour',
+});
+```
+
+#### `ActivityOptions`
+```ts
+// All timeouts and intervals accept ms format strings (see: https://www.npmjs.com/package/ms).
+
+// See: https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/activity/ActivityOptions.Builder.html
+interface CommonActivityOptions {
+  scheduleToCloseTimeout?: string,
+  startToCloseTimeout?: string,
+  scheduleToStartTimeout?: string,
+  heartbeatTimeout?: string,
+  /**
+   * If not defined, will not retry, otherwise retry with given options
+   */
+  retry?: RetryOptions,
+}
+
+interface LocalActivityOptions extends CommonActivityOptions {
+  type: 'local',
+}
+
+// 
+interface RemoteActivityOptions extends CommonActivityOptions {
+  type: 'remote',
+  taskQueue: string,
+}
+
+// See: https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/common/RetryOptions.Builder.html
+interface RetryOptions {
+  backoffCoefficient?: number,
+  initialInterval?: string,
+  maximumAttempts?: number,
+  maximumInterval?: string,
+}
+
+type ActivityOptions = RemoteActivityOptions | LocalActivityOptions;
+```
+
+
+### Signals
+Workflows can register signal hooks by via the exported `workflow`'s `signals` property.
+
+```ts
+const scope = new CancellationScope();
+
+export const signals = {
+  abandon(reason: string) {
+    console.log('Abandon requested, reason:', reason);
+    scope.cancel();
+  }
+}
+
+async function main() {
+  // ...
+}
+
+export const workflow: Example = { main, signals };
+```
+
+### Queries
+Workflows can register signal hooks by via the exported `workflow`'s `queries` property.
+
+```ts
+let step = 0;
+const state = {
+  a: 1,
+  b: 2,
+};
+
+export const queries = {
+  step: () => step,
+  state: (key) => state[key],
+}
+
+async function main() {
+  // ...
+  ++step;
+  state.b = 3;
+  // ...
+}
+
+export const workflow: Example = { main, signals };
+```
+
+### Limitations
+* The [NodeJS API][nodejs-api] is not available in workflow code as it breaks determinism.
+* Only ES modules can be imported out of the box with V8 isolates while most npm package use the CommonJS module implementation.
+    * We can bypass this by implementing `CommonJS` `require` ([reference](https://nodejs.org/api/modules.html#modules_all_together)).
+    * 3rd party imported modules may contain async / await keywords - this can be overcome by transpiling each required file at runtime.
+* Submitted workflow code must not use async / await syntax.
+    Using async / await bypasses the injected Promise object in the workflow context and hinders our ability to gain full control over scheduling.
```

Comment:
Is there a side-discussion around why it is necessary to inject a customized Promise object vs potentially greening some of the IO operations in the node API.  I'm just curious as I was recently looking into this for a discrete-event simulation system.

Reply:
@psigen I just recently figured out that we don't need to inject a customized Promise object.
Since we work directly with V8 and not the entire NodeJS runtime we can ensure that our workflow runtime is deterministic by controlling all async completions (e.g. activities and timers) and running all of the v8 microtasks each time we call back into the isolate.
I'll update the proposal soon in another PR.
Hope that answers your question.

Reply:
Thanks, that's exactly what I was hoping!